design/logging.md

```diff
@@ -0,0 +1,542 @@
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/method_override.rb:24:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/runtime.rb:22:in `call'
+/var/lib/gems/2.5.0/gems/activesupport-5.2.4.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
+/var/lib/gems/2.5.0/gems/actionpack-5.2.4.2/lib/action_dispatch/middleware/static.rb:127:in `call'
+/var/lib/gems/2.5.0/gems/rack-2.2.2/lib/rack/sendfile.rb:110:in `call'
+/var/lib/gems/2.5.0/gems/conjur-rack-heartbeat-2.2.0/lib/rack/heartbeat.rb:20:in `call'
+/var/lib/gems/2.5.0/gems/railties-5.2.4.2/lib/rails/engine.rb:524:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/configuration.rb:227:in `call'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:675:in `handle_request'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:476:in `process_client'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/server.rb:334:in `block in run'
+/var/lib/gems/2.5.0/gems/puma-3.12.4/lib/puma/thread_pool.rb:135:in `block in spawn_thread'
+ INFO 2020/05/18 09:45:41 +0000 [pid=298] [origin=127.0.0.1] [request_id=b7390d52-73a1-41ff-add3-9f4e37395e55] [tid=305] Completed 401 Unauthorized in 288ms
+```
+
+As you can see, this log shows too much data on why the request failed.
+This makes the log very hard to debug, and it's clear that we need some solution for this.
+
+## Solution
+
+According to Datadog's post on [managing Rails application logs](https://www.datadoghq.com/blog/managing-rails-application-logs/),
+the Rails log levels are defined as follows:
+- Debug: diagnostic information for developers and system administrators, including 
+  database calls or inspecting object attributes. This is the most verbose log level.
+- Info: useful information about normal application operations such as services starting or stopping.
+- Warn: operations that an application can easily recover from but should be addressed 
+  soon. This could include using an out-of-date gem or retrying an operation.
+- Error: errors that cause an operation to fail (e.g., missing data or files) but 
+  not the application. The issue should be resolved soon, but the application can log the exception and continue running.
+  
+We will go over our log messages and change their log level according to the above.
+DB calls will stay in DEBUG, whereas most of the logs that we write on our own (i.e `CONJ<some-number>D`)
+will be logged in INFO, and their codes will change accordingly (i.e `CONJ<some-number>D`).
+
+We will also change the codes for some of our error messages (i.e `CONJ<some-number>E`)
+as they will be written to the log in `INFO` level (today they are written in `DEBUG`). 
+The reason they should be logged in `INFO` and not in `ERROR`
+is that an `ERROR` log indicates `errors that cause an operation to fail (e.g., 
+missing data or files) but not the application`. In some of our error messages,
```

Comment:
Not critical but note that datadog/rails have also fatal which we don't use so I guess our Errors can also be Fatal

Reply:
yes, we don't have any log messages that crash our server.